urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,223 @@
+import errno
+import socket
+from urllib3.util.selectors import (
+    DefaultSelector,
+    EVENT_WRITE
+)
+from urllib3.util.timeout import current_time
+
+
+_HAPPY_EYEBALLS_CACHE = {}
+_HAPPY_EYEBALLS_CACHE_TIME = 60 * 10  # 10 minutes according to RFC 6555
+_ASYNC_ERROR_NUMBERS = set([errno.EINPROGRESS,
+                            errno.EAGAIN,
+                            errno.EWOULDBLOCK])
+if hasattr(errno, "WSAEWOULDBLOCK"):  # Platform-specific: Windows
+    _ASYNC_ERROR_NUMBERS.add(errno.WSAEWOULDBLOCK)
+
+_IP_FAMILIES = set([socket.AF_INET])
+if hasattr(socket, "AF_INET6"):
+    _IP_FAMILIES.add(socket.AF_INET6)
+
+
+class _HappyEyeballs(object):
+    def __init__(self, address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
+                 source_address=None, socket_options=None):
+        self._address = address
+        self._error = None
+        self._timeout = timeout
+        self._source_address = source_address
+        self._socket_options = socket_options
+        self._sockets = []
+        self._selector = DefaultSelector()
+        self._start_time = None
+
+    def connect(self):
+        """ Given the setup of the algorithm, attempt to
+        use a cached value and if it doesn't work then try
+        to use a """
+        self._start_time = current_time()
+        try:
+            sock = None
+            if self._address in _HAPPY_EYEBALLS_CACHE:
+                family, expires = _HAPPY_EYEBALLS_CACHE[self._address]
+
+                # Invalidate the cache if it's expired.
+                if expires > current_time():
+                    del _HAPPY_EYEBALLS_CACHE[self._address]
+                else:
+                    try:
+                        sock = self._algorithm(family)
+                    # Invalidate the cache if connecting doesn't work.
+                    except (OSError, socket.error):
+                        del _HAPPY_EYEBALLS_CACHE[self._address]
+
+            if sock is None:
+                sock = self._algorithm()
+
+                # Add a new entry to the cache if successful.
+                cache_entry = (sock.family,
+                               current_time() + _HAPPY_EYEBALLS_CACHE_TIME)
+                _HAPPY_EYEBALLS_CACHE[self._address] = cache_entry
+
+            return sock
+        finally:
+            self._cleanup()
+
+    def _algorithm(self, family=None):
+        """ Executes the algorithm for a given family
+        or no family if IPv4 or IPv6 are allowed. """
+        if family is None:
+            family = socket.AF_UNSPEC
```

Comment:
I don't think there's any reason to have the default of family be `None`: let's just set it to `socket.AF_UNSPEC` and save ourselves an if block.